larq_compute_engine/core/ruy_pack.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Ruy performs 'packing' of the LHS and RHS matrices before the GEMM operation,
+ * which involves re-arranging the matrices in memory to enable linear memory
+ * access patterns for the optimised assembly GEMM kernels. For more detail on
+ * the motivation for packing, see the the gemmlowp docs:
+ * https://github.com/google/gemmlowp/blob/master/doc/packing.md.
+ *
+ * Packing is part of the inference 'critical path' because although the packed
+ * weights matrix can be cached, the packed input activations cannot be. Ruy
+ * therefore includes optimised packing routines. However, these are specific to
+ * the datatypes (float and/or int8) and kernel-layouts used in the built-in Ruy
+ * kernels, which don't match those of our optimised LCE kernels. This means
+ * that if we rely on the Ruy packing code we end up falling back to the slow
+ * reference implementation that copies each element one-by-one.
+ *
+ * In our optimised LCE kernels, the LHS matrix is the weights and the RHS is
+ * the input activations. As the packed weights can be computed once then
+ * cached, we only care about optimising for packing the RHS. In all our
+ * optimised kernels, the RHS has a 4x4 column-major kernel layout; we therefore
+ * extend the Ruy packing with an optimised (but portable, C++) packing routine
+ * for 4x4 column-major kernel layouts.
+ */
+
+#include "larq_compute_engine/core/types.h"
+#include "ruy/pack_common.h"
+
+namespace compute_engine {
+namespace tflite {
+
+using compute_engine::core::TBitpacked;
+using namespace ruy;
+
+template <Path ThePath, typename FixedKernelLayout>
+struct LceRuyPackImpl
+    : PackImpl<ThePath, FixedKernelLayout, TBitpacked, TBitpacked, TBitpacked> {
+};
+
+template <Path ThePath>
+struct LceRuyPackImpl<ThePath, FixedKernelLayout<Order::kColMajor, 4, 4>> {
+  // This method is derived Ruy's 16x4 column-major int8 packing routine:
+  // https://github.com/google/ruy/blob/c9f5f9cecde3d6314df6e7d91517356bf07135eb/ruy/pack_arm.h#L122-L197
+  static void Run(Tuning, const Mat<TBitpacked>& src_matrix,
+                  PMat<TBitpacked>* packed_matrix, int start_col, int end_col) {
+    profiler::ScopeLabel label("Pack (ColMajor, 4x4)");
+
+    // Ruy supports collecting column sums during the packing process, which we
+    // have no need for.
+    RUY_DCHECK_EQ(packed_matrix->sums, nullptr);
+
+    // Likewise, Ruy supports arbitrary zero points, but we only use true-zero.
+    RUY_DCHECK_EQ(src_matrix.zero_point, 0);
+
+    RUY_DCHECK(IsColMajor(src_matrix.layout));
+    RUY_DCHECK(IsColMajor(packed_matrix->layout));
+    RUY_DCHECK_EQ(start_col % 4, 0);
+
+    TBitpacked zerobuf[4];
+    memset(zerobuf, 0, sizeof(zerobuf));
+
+    for (int block_col = start_col; block_col < end_col; block_col += 4) {
+      int src_stride = src_matrix.layout.stride;
```

Comment:
I don't think this matters much, but I guess it's nicer to move the assignment out of the loop:
```suggestion
    int src_stride = src_matrix.layout.stride;
    for (int block_col = start_col; block_col < end_col; block_col += 4) {
```